### PR TITLE
[python] Adding rowgroup evaluator using pyarrow parquet stats

### PR DESCRIPTION
**Pull request by TGooch44**
Adding evaluator for determining if a given pyarrow parquet row group may contain rows for the assigned predicate and read-range[start, end]

cc: @danielcweeks @rdblue

### CODE REVIEW

**Comment by rdblue** [comment]
@rymurr, you might be interested in reviewing this one.

**Comment by rymurr** [comment]
> @rymurr, you might be interested in reviewing this one.

Very interested! Please give me a day or two as a bit busy at the moment. 

I was looking at some similar stuff with the [Arrow Datasets](https://arrow.apache.org/docs/python/dataset.html) package.

**Comment by TGooch44** [comment]
I should have marked this is WIP. I'm still adding some logic to handle some of the other data types like timestamp.  I'll try to get this updated soon

**Comment by TGooch44** [comment]
@rymurr This is ready for review whenever you have a chance.  The failed check looks like it's something on the java side, which this PR isn't touching.

**Comment by rymurr** [comment]
Regarding Arrows dataset here is a **very** rough example of what I think we can do.

The filters from this PR, partition filtering, different filesystem types etc are handled here. Partitions by iceberg metadata and the rest by pyarrow. What do you guys think?

``` python
from iceberg.hive import HiveTables
import pyarrow as pa
import pyarrow.dataset as ds
from pyarrow import fs

if __name__ == '__main__':

    conf = {"hive.metastore.uris": 'thrift://localhost:9083',
            'hive.metastore.warehouse.dir': '/home/ryan/warehouse/iceberg/hive_test'}
    tables = HiveTables(conf)

    tbl = tables.load("testing.foo")

    # inspect metadata
    print(tbl.schema())
    print(tbl.spec())
    print(int(tbl.current_snapshot().summary.get("total-records")))

    scan = tbl.new_scan() \
        .filter("symbol==AUDCHF") \
        .select(["Bid", "Ask", "Datetime"])

    projection = scan.schema
    for task in scan.plan_tasks():
        dataset = ds.FileSystemDataset.from_paths([i.file._file_path.replace("file:","") for i in task.files],
                                                schema=pa.schema([("Bid", pa.float64()), ("Ask", pa.float64()),
                                                                  ("DateTime", pa.timestamp("us", 'UTC'))]),
                                                format=ds.ParquetFileFormat(),
                                                filesystem=fs.LocalFileSystem())
        pytbl = dataset.to_table(filter=ds.field("Bid") > 0.75)
        df = pytbl.to_pandas()
        print(df)
```

Output
```
table {
 1: DateTime: optional timestamptz(None)
 2: Bid: optional double(None)
 3: Ask: optional double(None)
 4: symbol: optional string(None)
}
[
 1000: DateTime_day: day(1)
]
846035
            Bid      Ask                         DateTime
0       0.75935  0.76156 2018-01-01 21:58:33.821000+00:00
1       0.75940  0.76155 2018-01-01 21:58:34.821000+00:00
2       0.75943  0.76154 2018-01-01 21:58:35.733000+00:00
3       0.75945  0.76153 2018-01-01 21:58:36.734000+00:00
4       0.75947  0.76152 2018-01-01 21:58:37.733000+00:00
...         ...      ...                              ...
388406  0.76612  0.76669 2018-01-05 21:59:00.070000+00:00
388407  0.76611  0.76669 2018-01-05 21:59:00.336000+00:00
388408  0.76611  0.76670 2018-01-05 21:59:00.809000+00:00
388409  0.76524  0.76747 2018-01-05 21:59:01.011000+00:00
388410  0.76534  0.76757 2018-01-05 21:59:12.367000+00:00

[388411 rows x 3 columns]
```

File: python/setup.py [hunk 1]
```diff
@@ -38,7 +37,7 @@
                       'requests',
                       'retrying',
                       'pandas',
-                      'pyarrow'
+                      'pyarrow>=0.17.0'
```
[the remark this answers is not in the document]
Comment (TGooch44) [reply python/setup.py:40]: 0.17.0 was the first release that contained field_id metadata necessary for schema evolution mapping.  I'm just coming back from being on leave, so I'm a little behind on what's changed from 0.17.0->1.0, I'll try to get caught up.

**Comment by TGooch44** [comment]
Let me review the additions to Dataset since it wasn't available when the bulk of our internal parquet code was written, the only reason for not more directly leveraging the arrow stuff was for enforcing schema evolution, eg. projecting nulls for columns that were added, renaming/reordering columns, etc.   

I agree that this would work a lot better if the heavy lifting is done in the arrow C libs.  I think with some tweaking we can probably make it happen.

File: python/setup.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rymurr) [reply python/setup.py:40]: welcome back!

**Comment by rymurr** [comment]
I doubt the Arrow Dataset code will have thought as long and as hard about schema evolution as you guys have! If you find any gaps in the dataset code I would love to help get the changes merged into Arrow

**Comment by rdblue** [comment]
I like using Arrow as much as possible. Like Ted said, the main concern is getting schema evolution right. Arrow/Parquet would need to implement ID-based schema evolution and at least some of the promotion rules like int -> long.

**Comment by TGooch44** [comment]
@rymurr I'm admittedly new to python type hints, so I took a crack at it, but would appreciate if you have a second to look it over.  Also, re: Arrow Datasets, would you be opening to jumping on a quick video chat?  I'd like to talk through it a bit with you if possible.

**Comment by rymurr** [comment]
> @rymurr I'm admittedly new to python type hints, so I took a crack at it, but would appreciate if you have a second to look it over. Also, re: Arrow Datasets, would you be opening to jumping on a quick video chat? I'd like to talk through it a bit with you if possible.

Hey @TGooch44 I am just knocking off for the day. I could do tomorrow your morning?

**Comment by TGooch44** [comment]
Next week may be better. I have twins that are still not really sleeping,
so my early mornings are a little touch and go.  I have night help
some nights, which makes planning morning meetings a lot easier.

Otherwise, depending on your flexibility schedule-wise there is a good
chance I'll be up sometime between 12am-6am PDT, and could take a meeting.
Hard to commit to a specific time though.

**Comment by rymurr** [comment]
Hey @TGooch44. Twins! :exploding_head:  Congrats and get some sleep! I am free from 11:30 GMT+1 if you are awake, else next week suits me perfectly fine.

**Pull request closed without merging**